File: Easy/002_PalindromePrime/C/prime.c
#include <stdio.h>
#include <stdlib.h>

void Sieve(int max, int *prime);
int FindPalindrome(int *primes, int len);

int main(int argc, char *argv[])
{
    int limit = 1001;
    int prime[limit];
    Sieve(limit, prime);
    
    printf("%d\n", FindPalindrome(prime, limit));
    
    return 0;
}

void Sieve (int max, int *prime)
{
    // set all to true
    for (int i=0; i<max+1; ++i)
        prime[i] = 1;
    
    // set 0 and 1 to false (not primes!)
    prime[0] = 0;
    prime[1] = 0;
    
    for (int i=2; i<max/2; ++i)
        for (int j=i*i; j<max; j += i)
            prime[j] = 0;
    
}

int FindPalindrome (int *primes, int len)
{
    int hi = 0;
    
    for (int i=0; i<len; ++i)
    {
            if (primes[i] == 1)
            {
            int n = i;
            int num = n;
            
            int reverse = 0;
            int digit = 0;
            
            while (num > 0)
            {
                digit = num % 10;
                
                reverse = reverse * 10 + digit;
                
                num /= 10;
            }
            
            if (n == reverse && n > hi)
                hi = n;
        }
            
    }
    
    return hi;
    
}

File: Easy/002_PalindromePrime/Python/prime.py
#!/usr/bin/env python3

'''
    This method creates a sieve of boolean
    values
'''
def Sieve(max):
    # init a list
    s = {}
    # loop through the list, setting all to true
    for i in range(0, max+1):
        s[i] = True
    
    # 0 & 1 are not primes
    s[0] = False
    s[1] = False
    
    # set the rest nonprimes to false
    for i in range(2, int(max/2)):
        for j in range(i*i, max, i):
            s[j] = False
    # return the sieve        
    return s

'''
    This method finds the highest palindrome in the
    set of prime numbers passed to method
'''
def Palindrome(primes, max):
    # var to hold highest
    hi = 0
    # loop through each in list
    for i in range(0, max+1):
        # if current is a prime number
        if primes[i] == True:
            # declares some variables
            n = i
            num = n
            reverse = 0
            digit = 0
            # while num is greater than 0, we still have digits
            # to pop off the back of the number
            while (num > 0):
                # pop the last digit off the number
                digit = int(num % 10)
                # add to the reverse
                reverse = int(reverse * 10 + digit)
                # divide the number by 10
                num = int(num / 10)
            # if the n is same as reverse and highest so far
            if n == reverse and n > hi:
                # set the hi
                hi = n
    # return the highest palindrome
    return hi

prime = Sieve(1000)
print (Palindrome(prime, 1000))
